### DISCUSSION TITLE
Do you guys support DUO in any way?

### DISCUSSION THREAD

**Discussion by gdonval** [kanidm/kanidm#3275] [Q&A] [answered]
We're currently using `pam_duo` + `ldap/sssd` stack for authentication. Duo provides a ldap proxy that does the Duo request on the user's behalf and that would simplify things tremendously. The proxy code is really nasty though (in itself and in the way dependencies are "managed"). I was wondering if you guys had experience with using Kanidm with Duo, especially in a way that does not require each single machine to be "Duo-aware".

**Comment by yaleman** [comment]
There's no integration that has been tested/written, and I'm not sure how that'd even look,where would you want Kanidm to fit in?

**Reply by gdonval** [reply]
> where would you want Kanidm to fit in?

In my mind, with my constraints (which I'm aware are not yours), I would love a system that stands on its own (like Kanidm already does), can use an existing LDAP as an authentication source (among others) and would allow me to hook [`login_duo`](https://duo.com/docs/loginduo) in so that whenever someone using a legacy-LDAP-sourced account, a push notification would be sent onto their phone, transparently. Kanidm would only send the "OK" after the push notification is cleared in "OK" state.

Duo provides a LDAP-proxy that does something along those lines. It's just a nightmare to sysadmin and it looks brittle.

**Reply by gdonval** [reply]
I know Kanidm already does MFA and is meant to be its own DB but my current constraints are that I must use our existing LDAP-based backend for auth, including supporting the Duo layer. I try to see where Kanidm fits within those constraints.

**Comment by Firstyear** [comment, marked as answer]
We don't integrate with duo, no - duo is a "bolt on mfa" for legacy auth systems. Kanidm unlike LDAP has native MFA and so doesn't require bolt-ons like duo.

**Reply by gdonval** [reply]
Thanks for the answer: it clarifies your vision for Kanidm, which help a lot.

I don't intend to really make a plea for it because I understand your point but Duo bring something on the table that, I believe, you don't provide: authentication through an app through push notifications. It's not as good as passwordless and not as privacy-respecting as OTP but it's a rather nice middleground from an organization's perspective.

So I would love for it to be supported but I understand it doesn't align with the project's priorities and philosophy.

**Reply by Firstyear** [reply]
I think it's only a middleground for an org *because* passwordless wasn't on the cards previously. Now that it's available, it's a better experience generally than a push based app.
